### ISSUE TITLE
PHP User Auth part duex...

### ISSUE THREAD

**Issue by lookfirst** [kreait/firebase-php#312]
Re-opennig because #159 is closed to comments.

There is a 4th case to https://github.com/kreait/firebase-php/issues/159#issuecomment-360225655 that you didn't consider in your comment.

`backend to backend`

I want to give access to my backend API, which is secured through Firebase Admin http header Authorization tokens with verifyToken(). I can just allow the other backend to create an email/password account and then they can deal with securing that e/p themselves.

This makes it so I don't have to give them a service account credential, just a user/password account that they can create themselves and I can easily add in a custom claim on their account to further lock down their account access.

**Comment by jeromegamez** [comment]
Old issues get closed and locked automatically because what's discussed in them might not be true/false anymore at the current date. This is the case here as well.

If you have ensured that a user has authenticated correctly, you can impersonate them by using the following methods as documented:

- [User management: Get information about a specific user](https://firebase-php.readthedocs.io/en/latest/user-management.html#get-information-about-a-specific-user)
- [Authentication: Authenticate with limited privileges](https://firebase-php.readthedocs.io/en/latest/authentication.html#authenticate-with-limited-privileges)

So, if I understand your use case correctly, something like this should get you started:

```php
$factory = (new Factory())->withServiceAccount();

$adminConnection = $factory->create();

$userEmail = '...';
$optionalClaims = [/* ... */];

$mappedUser = $adminConnection->getAuth()->getUserByEmail($userEmail);

$userConnection = $factory->asUser($mappedUser->uid, $optionalClaims)->create();

$userDatabase = $userConnection->getDatabase();
```

I hope this helps!

**Comment by lookfirst** [comment]
No worries about auto closing issues.

I am sorry, but I think you're still not understanding the use case. Let me try to explain again.

I have a backend api. It is secured by passing in a token through an Authorization header. FirebaseAuth.verifyIdToken() makes sure the token is good.

What I want to offer to a third party, who also runs a backend service (using PHP), a way to securely call my api but from their backend service.

In order to give them access, I have 3 choices:

1. Give them a special service account json file to auth with.
2. Give them an api where they can send a e/p and generate a custom token which they can then use as an idToken.
3. Use a library like yours to generate an idToken themselves since they store their own e/p themselves.

This is the use case that you didn't consider in the last issue and it is why it is important to offer the equivalent of firebase.auth().signInWithEmailAndPassword(), as a PHP option.

**Comment by jeromegamez** [comment]
Thanks for taking the time to clear this up. Unfortunately, I still stand by the points that I made in in #159 with the following additions:

Concerning Option 2: providing a custom token to a third party will not be sufficient: You can't use a custom token _as_ an iDToken, you have to use it to _exchange_ it with an ID Token, and you can do this only with a request authenticated by a ServiceAccount with admin privileges.

For both options 2 and 3, you would have to provide the third party with an Admin ServiceAccount (which the official and this unofficial Admin SDKs need) - both of which I don't think is desirable.

If you want that third parties can access your Firebase Project through your own API, you have to proxy the requests.

If you want them to be able to access your Firebase Project directly, they need a Service Account.

I'd like to point out that `signIn*()` methods are only available in Client SDKs, not in Admin SDKs, see for example the [Admin Node SDK](https://firebase.google.com/docs/reference/admin/node/admin.auth.Auth.html) or the [Admin Java SDK](https://firebase.google.com/docs/reference/admin/java/reference/com/google/firebase/auth/FirebaseAuth)

What I did notice while looking through the docs is that there seems to be a `generateSignInWithEmailLink()` method in the official Admin SDKs that I either missed or that was added since I looked last time 😅 - perhaps this can help somehow, and I will make sure to include that in one of the upcoming releases.

Please let me know if you think I still got things wrong, my intention is not to "shut you down" - I still think it's very unlikely that I will include this kind of functionality in the SDK, but I'm interested in your thoughts nonetheless!

**Comment by jeromegamez** [comment]
Just to be clear: I think it's perfectly fine to create an API endpoint where third parties can sign in with email and password and get an ID token in return.

They can then use the ID token to authenticate their following requests to your API, where you can verify the ID Token to ensure that the token is still valid.

That is basically the implementation of the `signInWithEmailAndPassword()` pattern, and in this case, the third party backend is "the user" your API application is "the Client SDK" and the Admin SDK stays the Admin SDK 😅

**Comment by lookfirst** [comment]
Thank you for the detailed response. Appreciate the discourse! You've been great.

I think now that you are considering ONLY the Admin SDK. Everything you think about revolves around a serviceAccount, which is of course necessary for the Admin SDK to do anything. I really dislike their naming scheme, which I kind of blame for this confusion. There is no 'client' SDK, it is a 'web' SDK and Admin isn't really 'admin', it is more like 'Service Account'.

Another issue is that the whole Firebase usecase is basically being a backend for a mobile app. They don't really have the case where the provider of that app might want to make their own API available and still take advantage of their authentication api's. In the OAuth world, it would be to make an Applications account, similar to what Github has in their ['Applications'](https://github.com/settings/installations) functionality. I just submitted a feature request to Firebase for this, so we will see.

The `generateSignInWithEmailLink` method works great if it is a real person. But I'm talking about two computers talking to each other and authenticating over email that way doesn't seem very efficient. ;-)

What I'm asking for here, and I apologize, really indirectly, so I will make it more directly, is that you implement the few auth methods in the [Client SDK](https://firebase.google.com/docs/auth/web/start) as well. But really, I only need one method... `signInWithEmailAndPassword`. Since you already have a lot of the code around creating the FirebaseUser and such, it shouldn't be too much effort.

Yes, I can provide an endpoint, but that is more code for me to maintain. I prefer to maintain less code by nature as that keeps things simpler. ;-)

So yea... since that doesn't seem interesting to you, it is fine... I'll go looking to see if someone else has implemented it or maybe implement it myself. My point though, in opening this issue was more to maybe just enlighten you to another method for authentication that you didn't consider.

Once again, super appreciate it!

**Comment by jeromegamez** [comment]
> Yes, I can provide an endpoint, but that is more code for me to maintain. I prefer to maintain less code by nature as that keeps things simpler. ;-)

That made me chuckle because my first reaction was: „Eh, that‘s actually my argument!“ 😂

I remember that I thought about all this for a long time in the 2.x/3.x era of this SDK, and that I had that feature once earlier, but then removed it because of good reasons... that I don‘t remember 😅

I‘ll keep this issue open, let it sink in and revisit it when I have some headspace, but know that I hear you.

I have another question: when you sign in a user like this, you will create state (the user is now signed in, and its user record in the Firebase Auth database will be changed with the last sign in date). Will you then pass the idToken that the Firebase API returns back to the third party, they use it on their next request to your API where you would then have to verify it again? Or do they authenticate each request with username and password and you newly sign them in each time?

**Comment by jeromegamez** [comment]
I think I‘m still switching constantly between understanding the use case and not understanding the use case 😅

I think I should restart simple: are you intending the 3rd party to use the SDK in their application as well and they use only a `signInWithEmailAndPassword()` method to then authenticate requests to a given Firebase project themselves?

Sorry for the back and forth 😅

**Comment by jeromegamez** [comment]
I think I'm getting somewhere - each Firebase project has a Web API key on (https://console.firebase.google.com/u/1/project/_/settings/general/) that can be used to authenticate requests to the Firebase Auth REST API, so if you provide the 3rd party backend with the Web API key, they can generate an ID token without needing to install this SDK or needing to access your API to do so. Here's a proof of concept:

```php
<?php

require 'vendor/autoload.php';

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use Psr\Http\Message\ResponseInterface;

$apiKey = '...';

$client = new Client([
    'http_errors' => false,
]);

$headers = ['Content-Type' => 'application/json'];

$signUpRequest = static function ($email, $password) use ($apiKey, $headers): Request {
    $url = 'https://identitytoolkit.googleapis.com/v1/accounts:signUp?key='.$apiKey;
    $body = json_encode(['email' => $email, 'password' => $password, 'returnSecureToken' => true]);

    return new Request('POST', $url, $headers, $body);
};

$signInRequest = static function ($email, $password) use ($apiKey, $headers): Request {
    $url = 'https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key='.$apiKey;
    $body = json_encode(['email' => $email, 'password' => $password, 'returnSecureToken' => true]);

    return new Request('POST', $url, $headers, $body);
};

$printResponseData = static function (ResponseInterface $response)
{
    echo json_encode(json_decode((string) $response->getBody(), true), JSON_PRETTY_PRINT);
};

$getIdTokenFromResponse = static function (ResponseInterface $response): string
{
    return json_decode((string) $response->getBody(), true)['idToken'];
};

$email = 'jerome@domain.tld';
$password = 'mysecurepassword';

// Signing up only works the first time
// $printResponseData($client->send($signUpRequest($email, $password)));

echo $getIdTokenFromResponse($client->send($signInRequest($email, $password)));
```

The only downside I see here is that giving away the Web-API-Key enables the wielder to sign up new accounts at will - but that's also the case for Web Applications where the key is also exposed and seems to be a known (non-)problem.

What do you think?

**Comment by lookfirst** [comment]
Yes, you're getting it! 

Again, the idea is that we want to allow a computer to sign in without any human interaction. Once they have the idToken from signing in, they can then hit my cloud functions which call verifyIDToken.

For now, I just implemented this as a cloud function on my backend... they can hit my /token endpoint with an e/p (over https) and get back a token. No need to hand out my web api key. But now that you mention that that is needed, I can see that just providing a /token endpoint is the best solution here.

So... despite more code, I implemented that yesterday. Then they can hit any other cloud function endpoint that I offer up if they pass the token in with an 'authorized: bearer IDTOKEN' http header.

```
/* tslint:disable:no-import-side-effect */
import 'firebase/auth';
import * as clientApp from "firebase/app";

import {Request, Response, Router} from "express";

/**
 * API endpoint to generate a token which can be used to talk to the backend API.
 */
export const generateToken = (router: Router, firebaseClientApp: clientApp.app.App) => {
	return router.post('/token', async (req: Request, res: Response) => {
		try {
			const email = req.body.email;
			const password = req.body.password;
			const userCredential = await firebaseClientApp.auth().signInWithEmailAndPassword(email, password);
			if (userCredential && userCredential.user) {
				const idTokenResult = await userCredential.user.getIdToken(true);
				res.json({status: 'success', token: idTokenResult});
			}
		} catch (error) {
			console.log(error);
			res.json(error);
		}
	});
};
```

Called with the token string from the http header:

```
export async function verifyIdToken(firebaseAdminApp: adminApp.app.App, idToken: string): Promise<string> {
	const decodedToken = await firebaseAdminApp.auth().verifyIdToken(idToken);
	const userRecord = await firebaseAdminApp.auth().getUser(decodedToken.uid);

	if (userRecord && userRecord.customClaims) {
		if (!userRecord.customClaims.hasOwnProperty('enabled')) {
			throw {
				errorInfo: {
					code: 'User does not have the enabled custom claim.'
				}
			};
		}
	}

	return decodedToken.uid;
}
```

I run this as a middleware and use an 'enabled' custom claim flag to weed out accounts that, as you can say, anyone can generate. I have another endpoint which 'enables' their account. I may change that in the future to something more descriptive, but 'enabled' is simple enough for now...

```
import {Request, Response, Router} from "express";
import {app as adminApp} from "firebase-admin";
import {METADATA} from "../auth";

/**
 * Enabled allows us to set a custom user class property on a user by
 * an email address. We use this property to restrict access to the scanner app.
 *
 * This ties in with another cloud function which gets run every
 * time a user is created or deleted.
 */
export const enabledApi = (router: Router, firebaseAdminApp: adminApp.App) => {
	return router.post('/enabled', async (req: Request, res: Response) => {
		const email = req.body.email;
		const enabled = req.body.enabled;

		try {
			// Set the enabled claim on the user
			const user = await firebaseAdminApp.auth().getUserByEmail(email);
			await firebaseAdminApp.auth().setCustomUserClaims(user.uid, {enabled: enabled});

			// Update real-time database to notify client to force refresh.
			const metadataRef = firebaseAdminApp.database().ref(METADATA + user.uid);
			await metadataRef.set({refreshTime: new Date().getTime()});

			res.json({status: 'success', enabled: enabled});
		} catch (error) {
			console.log(error);
			res.json(error);
		}
	});
};
```

**Comment by jeromegamez** [comment]
Great work, thank you for sharing the code, this will certainly help people who stumble upon this thread to learn how this is done with cloud functions!

But do I understand this right that this would have been much easier if I had just™ implemented the `signInWithEmailAndPassword()` in the SDK? Because after having learned the use case by our conversations, I now feel inclined to add this :)

**Comment by lookfirst** [comment]
I definitely think it is still a good addition since it will pair with people who use PHP.

The only downside is that you have to give out your web key, which is something that I had not considered before. In my case, I only have a private flutter iphone/android app as public client where accounts can be created, so excessive account creation and key exposure, is not an issue for me.

The solution above for creating the `/token` endpoint is probably the end solution for me though because of the key issue.

I've been thinking separately that it wouldn't be too hard to create some UX around an OAuth application provider as well. This way people could self service their own tokens without having to create user accounts. But that would be a totally separate thing.

**Comment by thaifani** [comment]
> Great work, thank you for sharing the code, this will certainly help people who stumble upon this thread to learn how this is done with cloud functions! 🥰
> 
> But do I understand this right that this would have been much easier if I had just™ implemented the `signInWithEmailAndPassword()` in the SDK? Because after having learned the use case by our conversations, I now feel inclined to add this :)
please re implement the  signInWithEmailAndPassword() as you said in the 
https://github.com/kreait/firebase-php/issues/312#issuecomment-520275305

**Comment by jeromegamez** [comment]
Please check #390 for the current state of implementation.

**Comment by jeromegamez** [comment]
Released with [4.41.0](https://github.com/kreait/firebase-php/releases/tag/4.41.0) 🚀
